File: src/java/pulse/cdm/patient/actions/SERespiratoryFatigue.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package pulse.cdm.patient.actions;

import pulse.cdm.bind.PatientActions.RespiratoryFatigueData;
import pulse.cdm.properties.SEScalar0To1;

public class SERespiratoryFatigue extends SEPatientAction
{

  private static final long serialVersionUID = -8967263581748846332L;
  protected SEScalar0To1 severity;
  
  public SERespiratoryFatigue()
  {
    severity = null;
  }
  
  @Override
  public void reset()
  {
    super.reset();
    if (severity != null)
      severity.invalidate();
  }
  
  public void copy(SERespiratoryFatigue other)
  {
    if(this==other)
      return;
    super.copy(other);
    if (other.severity != null)
      getSeverity().set(other.getSeverity());
    else if (severity != null)
      severity.invalidate();
  }
  
  @Override
  public boolean isValid()
  {
    return hasSeverity();
  }
  
  public static void load(RespiratoryFatigueData src, SERespiratoryFatigue dst) 
  {
    SEPatientAction.load(src.getPatientAction(), dst);
    if(src.hasSeverity())
      SEScalar0To1.load(src.getSeverity(),dst.getSeverity());
  }
  
  public static RespiratoryFatigueData unload(SERespiratoryFatigue src)
  {
    RespiratoryFatigueData.Builder dst = RespiratoryFatigueData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  
  protected static void unload(SERespiratoryFatigue src, RespiratoryFatigueData.Builder dst)
  {
    SEPatientAction.unload(src,dst.getPatientActionBuilder());
    if (src.hasSeverity())
      dst.setSeverity(SEScalar0To1.unload(src.severity));
  }
  
  public boolean hasSeverity()
  {
    return severity == null ? false : severity.isValid();
  }
  public SEScalar0To1 getSeverity()
  {
    if (severity == null)
      severity = new SEScalar0To1();
    return severity;
  }

  @Override
  public String toString()
  {
    if (severity != null)
      return "Respiratory Fatigue" 
          + "\n\tSeverity: " + getSeverity();
    else
      return "Action not specified properly";
  }
}

File: src/cpp/cdm/patient/actions/SERespiratoryFatigue.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "patient/actions/SERespiratoryFatigue.h"
#include "properties/SEScalar0To1.h"
#include "io/protobuf/PBPatientActions.h"

SERespiratoryFatigue::SERespiratoryFatigue(Logger* logger) : SEPatientAction(logger)
{
  m_Severity=nullptr;
}

SERespiratoryFatigue::~SERespiratoryFatigue()
{
  SAFE_DELETE(m_Severity);
}

void SERespiratoryFatigue::Clear()
{
  SEPatientAction::Clear();
  INVALIDATE_PROPERTY(m_Severity);
}

void SERespiratoryFatigue::Copy(const SERespiratoryFatigue& src, bool preserveState)
{
  //if(preserveState) // Cache any state before copy,
  PBPatientAction::Copy(src, *this);
  //if(preserveState) // Put back any state
}

bool SERespiratoryFatigue::IsValid() const
{
  return SEPatientAction::IsValid() && HasSeverity();
}

bool SERespiratoryFatigue::IsActive() const
{
  if (!SEPatientAction::IsActive())
    return false;
  return !m_Severity->IsZero();
}
void SERespiratoryFatigue::Deactivate()
{
  SEPatientAction::Deactivate();
  Clear();//No stateful properties
}

const SEScalar* SERespiratoryFatigue::GetScalar(const std::string& name)
{
  if (name.compare("Severity") == 0)
    return &GetSeverity();
  return nullptr;
}

bool SERespiratoryFatigue::HasSeverity() const
{
  return m_Severity==nullptr?false:m_Severity->IsValid();
}
SEScalar0To1& SERespiratoryFatigue::GetSeverity()
{
  if(m_Severity==nullptr)
    m_Severity=new SEScalar0To1();
  return *m_Severity;
}
double SERespiratoryFatigue::GetSeverity() const
{
  if (m_Severity == nullptr)
    return SEScalar::dNaN();
  return m_Severity->GetValue();
}

void SERespiratoryFatigue::ToString(std::ostream &str) const
{
  str << "Patient Action : Respiratory Fatigue"; 
  if(HasComment())
    str<<"\n\tComment: "<<m_Comment;
  str << "\n\tSeverity: "; HasSeverity() ? str << *m_Severity : str << "Not Set";
  str << std::flush;
}